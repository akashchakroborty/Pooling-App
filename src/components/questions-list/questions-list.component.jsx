import React, { useEffect } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import Question from "../question/question.component";

import { QuestionsHeader, QuestionsContainer } from "./questions-list.styles";

import QuestionsListActionTypes from "../../redux/questions-list/questions-list.types";
import { createStructuredSelector } from "reselect";
import {
  selectQuestions,
  selectIsQuestionsListFetching
} from "../../redux/questions-list/questions-list.selector";

const QuestionsList = ({ dispatch, questions, isFetching }) => {
  useEffect(() => {
    dispatch({
      type: QuestionsListActionTypes.FETCH_QUESTIONSLIST_START
    });
  }, [dispatch]);

  return (
    <div>
      <QuestionsHeader>
        <h1>Questions</h1>
      </QuestionsHeader>
      <QuestionsContainer>
        {isFetching ? (
          <h1>Loading...</h1>
        ) : (
          questions.map((question, index) => (
            <Question
              key={index}
              question={question.question}
              published_at={question.published_at}
              choices={question.choices}
              url={question.url}
            ></Question>
          ))
        )}
      </QuestionsContainer>
    </div>
  );
};

QuestionsList.propTypes = {
  isFetching: PropTypes.bool.isRequired,
  questions: PropTypes.array.isRequired,
  dispatch: PropTypes.func.isRequired
};

const mapStateToProps = createStructuredSelector({
  questions: selectQuestions,
  isFetching: selectIsQuestionsListFetching
});

export default connect(mapStateToProps)(QuestionsList);
