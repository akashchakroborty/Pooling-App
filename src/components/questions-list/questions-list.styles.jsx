import styled from "styled-components";

export const QuestionsHeader = styled.div`
  text-align: center;
  background: #0d223f;
  max-width: 80vw;
  width: 80%;
  margin: 0 auto;
  border-top: 10px solid #ffc627;
  height: 100px;
  padding-top: 12px;
`;

export const QuestionsContainer = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-column-gap: 20px;
  grid-row-gap: 20px;
  justify-items: stretch;
  align-items: stretch;
  width: 80%;
  margin: 1rem auto;
  @media only screen and (max-width: 760px) {
    grid-template-columns: 1fr 1fr;
  }
  @media only screen and (max-width: 480px) {
    grid-template-columns: 1fr;
  }
`;
