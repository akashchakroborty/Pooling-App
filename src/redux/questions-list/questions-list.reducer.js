import QuestionsListActionTypes from "./questions-list.types";
const INITIAL_STATE = {
  questions: [],
  isFetching: false,
  errorMessage: undefined
};

const questionsListReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case QuestionsListActionTypes.FETCH_QUESTIONSLIST_START:
      return {
        ...state,
        isFetching: true
      };
    case QuestionsListActionTypes.FETCH_QUESTIONSLIST_SUCCESS:
      return {
        ...state,
        isFetching: false,
        questions: action.payload
      };
    case QuestionsListActionTypes.FETCH_QUESTIONSLIST_FAILURE:
      return {
        ...state,
        isFetching: false,
        errorMessage: action.payload
      };
    default:
      return state;
  }
};

export default questionsListReducer;
