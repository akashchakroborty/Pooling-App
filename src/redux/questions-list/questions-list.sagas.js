import { call, put, takeLatest, all, delay } from "redux-saga/effects";
import QuestionListActionTypes from "./questions-list.types";
import axios from "axios";

const ROOT_URL = "https://polls.apiblueprint.org";

function poolQuestionListApi() {
  return axios.request({
    method: "get",
    url: `${ROOT_URL}/questions`
  });
}

function addnewQuestionApi(questionData) {
  return axios.request({
    method: "post",
    url: `${ROOT_URL}/questions?page=1`,
    data: JSON.stringify(questionData)
  });
}

export function* fetchQuestionsListAsync() {
  try {
    let { data } = yield call(poolQuestionListApi);
    yield put({
      type: QuestionListActionTypes.FETCH_QUESTIONSLIST_SUCCESS,
      payload: data
    });
  } catch (error) {
    yield put({
      type: QuestionListActionTypes.FETCH_QUESTIONSLIST_FAILURE,
      payload: error.message
    });
  }
}

export function* addNewQuestionAsync({ payload }) {
  console.log(payload);
  try {
    yield call(addnewQuestionApi, payload.questionData);
    yield put({
      type: QuestionListActionTypes.ADD_NEW_QUESTION_SUCCESS,
      payload: "success"
    });
    yield call(payload.callback);
    yield delay(5000);
    yield put({
      type: QuestionListActionTypes.MESSAGE_HIDE,
      payload: false
    });
  } catch (error) {
    yield put({
      type: QuestionListActionTypes.ADD_NEW_QUESTION_FAILURE,
      payload: "failure"
    });
    yield delay(5000);
    yield put({
      type: QuestionListActionTypes.MESSAGE_HIDE,
      payload: false
    });
  }
}

export function* fetchQuestionsListStart() {
  yield takeLatest(
    QuestionListActionTypes.FETCH_QUESTIONSLIST_START,
    fetchQuestionsListAsync
  );
}

export function* addNewQuestionStart() {
  yield takeLatest(
    QuestionListActionTypes.ADD_NEW_QUESTION_START,
    addNewQuestionAsync
  );
}

export function* questionsListSagas() {
  yield all([call(fetchQuestionsListStart), call(addNewQuestionStart)]);
}
