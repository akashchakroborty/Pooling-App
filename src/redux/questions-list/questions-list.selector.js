import { createSelector } from "reselect";

const selectquestionsList = state => state.questionsList;

export const selectQuestions = createSelector(
  [selectquestionsList],
  questionsList => questionsList.questions
);

export const selectIsQuestionsListFetching = createSelector(
  [selectquestionsList],
  questionsList => questionsList.isFetching
);
