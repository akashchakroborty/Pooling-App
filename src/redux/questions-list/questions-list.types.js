const QuestionsListActionTypes = {
  FETCH_QUESTIONSLIST_START: "FETCH_QUESTIONSLIST_START",
  FETCH_QUESTIONSLIST_SUCCESS: "FETCH_QUESTIONSLIST_SUCCESS",
  FETCH_QUESTIONSLIST_FAILURE: "FETCH_QUESTIONSLIST_FAILURE",
  ADD_NEW_QUESTION_START: "ADD_NEW_QUESTION_START",
  ADD_NEW_QUESTION_SUCCESS: "ADD_NEW_QUESTION_SUCCESS",
  ADD_NEW_QUESTION_FAILURE: "ADD_NEW_QUESTION_FAILURE",
  MESSAGE_HIDE: "MESSAGE_HIDE"
};

export default QuestionsListActionTypes;
