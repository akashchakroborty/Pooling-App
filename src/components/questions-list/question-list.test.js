import React from "react";
import { shallow } from "enzyme";
import { QuestionsList } from "./questions-list.component";
import Question from "../question/question.component";

describe("QuestionsList component", () => {
  let mockDispatch;
  let mockQuestions;
  let mockIsFetching;
  let wrapper;

  beforeEach(() => {
    mockDispatch = jest.fn();
    mockQuestions = [
      {
        question: "mock",
        published_at: "mock",
        choices: [],
        url: "mock"
      }
    ];
    mockIsFetching = false;

    const mockProps = {
      dispatch: mockDispatch,
      questions: mockQuestions,
      isFetching: mockIsFetching
    };

    wrapper = shallow(<QuestionsList {...mockProps} />);
  });

  it("should render QuestionsList component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render an equal number of Question components as the questions prop", () => {
    expect(wrapper.find(Question).length).toEqual(mockQuestions.length);
  });

  it("should render Loading if isFetching is true", () => {
    const mockProps = {
      questions: [],
      isFetching: true,
      dispatch: mockDispatch
    };

    const newWrapper = shallow(<QuestionsList {...mockProps} />);
    expect(newWrapper.exists(".loading")).toBe(true);
  });
});
