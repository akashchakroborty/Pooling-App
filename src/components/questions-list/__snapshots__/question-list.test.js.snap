// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionsList component should render QuestionsList component 1`] = `ShallowWrapper {}`;
